File: web/src/app/services/ride.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import {PanicRequest, PassengerRide, RideRequest, StopRideRequest, StopRideResponse} from '../model/ride';
import { RideHistoryResponse, RideDetailResponse, RideRatingRequest, PassengerRideHistoryResponse } from '../model/ride-history';
import { environment } from '../../environments/environment';
import {
  RideTrackingResponse,
  InconsistencyReportRequest,
  PanicRideRequest
} from '../model/ride-tracking';


@Injectable({
  providedIn: 'root'
})
export class RideService {
  readonly API_URL = `${environment.apiUrl}/rides`;

  constructor(private http: HttpClient) {}

  // Get driver's ride history with optional date filtering
  getDriverHistory(
    startDate?: string,
    endDate?: string,
    sortBy: string = 'createdAt',
    sortDirection: string = 'DESC'
  ): Observable<RideHistoryResponse[]> {
    let params = new HttpParams()
      .set('sortBy', sortBy)
      .set('sortDirection', sortDirection);

    if (startDate) {
      params = params.set('startDate', startDate);
    }
    if (endDate) {
      params = params.set('endDate', endDate);
    }

    return this.http.get<RideHistoryResponse[]>(`${this.API_URL}/history/driver`, { params });
  }

  getPassengerHistory(
    startDate?: string,
    endDate?: string,
    page: number = 0,
    size: number = 10,
    sort: string = 'createdAt,desc',
  ): Observable<PassengerRide[]> {

    let params = new HttpParams()
      .set('page', page)
      .set('size', size)
      .set('sort', sort)

    if (startDate) {
      params = params.set('startDate', startDate);
    }
    if (endDate) {
      params = params.set('endDate', endDate);
    }

    return this.http.get<PassengerRide[]>(`${this.API_URL}/passenger`, { params });
  }

  // Get detailed information for a specific ride (for driver)
  getDriverRideDetail(rideId: number): Observable<RideDetailResponse> {
    return this.http.get<RideDetailResponse>(`${this.API_URL}/${rideId}/details/driver`);
  }

  // Get detailed information for a specific ride (for passenger)
  getPassengerRideDetail(rideId: number): Observable<RideDetailResponse> {
    return this.http.get<RideDetailResponse>(`${this.API_URL}/${rideId}/details/passenger`);
  }

  // Track a ride in real-time - get current status, location, and estimates
  trackRide(rideId: number): Observable<RideTrackingResponse> {
    return this.http.get<RideTrackingResponse>(`${this.API_URL}/${rideId}/track`);
  }

  // Update passenger location during ride
  updatePassengerLocation(rideId: number, location: { latitude: number; longitude: number }): Observable<void> {
    return this.http.put<void>(`${this.API_URL}/${rideId}/passenger-location`, location);
  }

  // Update vehicle location during ride (for testing/simulation)
  updateVehicleLocation(rideId: number, location: { latitude: number; longitude: number }): Observable<void> {
    return this.http.put<void>(`${this.API_URL}/${rideId}/vehicle-location`, location);
  }

  // Rate a completed ride
  rateRide(rideId: number, request: RideRatingRequest): Observable<void> {
    return this.http.post<void>(`${this.API_URL}/${rideId}/rate`, request);
  }

  // Trigger panic alert for a ride
  ridePanic(panicRequest: PanicRideRequest): Observable<void> {
    return this.http.put<void>(`${this.API_URL}/panic`,  { panicRequest: panicRequest });
  }

  // Cancel a ride with a reason
  cancelRide(reason: string): Observable<void> {
    return this.http.put<void>(`${this.API_URL}/cancel`, { reason: reason });
  }

  bookRide(orderData: any): Observable<any> {
    return this.http.post(`${this.API_URL}/order`, orderData, {
      responseType: 'text'
    });
  }

  // Use this method on 'Date' object before sending to backend
  // if backend is expecting 'LocalDateTime'
  formatLocalDateTime(date: Date | null): string | null {
    if (!date) return null;

    const pad = (n: number) => n < 10 ? '0' + n : n;

    return date.getFullYear() + '-' +
      pad(date.getMonth() + 1) + '-' +
      pad(date.getDate()) + 'T' +
      pad(date.getHours()) + ':' +
      pad(date.getMinutes()) + ':' +
      pad(date.getSeconds());
  }
  stopRide(stopRideRequest: StopRideRequest) : Observable<StopRideResponse> {
    return this.http.put<StopRideResponse>(`${this.API_URL}/stop`, stopRideRequest);
  }

  completeRide(rideId: number): Observable<void> {
    return this.http.put<void>(`${this.API_URL}/${rideId}/complete`, {});
  }

  reportInconsistency(rideId: number, request: InconsistencyReportRequest): Observable<void> {
    return this.http.post<void>(`${this.API_URL}/${rideId}/report-inconsistency`, request);
  }

  startRide(rideId: number): Observable<any> {
    return this.http.put(`${this.API_URL}/${rideId}/start`, {});
  }
}

File: mobile/app/src/main/java/com/example/mobile/network/service/RideService.java
package com.example.mobile.network.service;

import com.example.mobile.dto.ride.CancellationRequest;
import com.example.mobile.dto.ride.DriverRideDetailResponse;
import com.example.mobile.dto.ride.DriverRideHistoryResponse;
import com.example.mobile.dto.ride.InconsistencyReportRequest;
import com.example.mobile.dto.ride.PanicRideRequest;
import com.example.mobile.dto.ride.PassengerRideDetailResponse;
import com.example.mobile.dto.ride.PassengerRideHistoryResponse;
import com.example.mobile.dto.ride.RideRatingRequest;
import com.example.mobile.dto.ride.RideRatingStatusResponse;
import com.example.mobile.dto.ride.RideTrackingResponse;
import com.example.mobile.dto.ride.StopRideRequest;
import com.example.mobile.dto.ride.StopRideResponse;
import com.example.mobile.dto.vehicle.UpdateVehicleLocationRequest;

import java.util.List;

import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface RideService {

    @GET("api/v1/rides/history/driver")
    Call<List<DriverRideHistoryResponse>> getDriverRideHistory(
            @Query("startDate") String startDate,
            @Query("endDate") String endDate,
            @Query("sortBy") String sortBy,
            @Query("sortDirection") String sortDirection
    );

    @GET("api/v1/rides/{rideId}/details/driver")
    Call<DriverRideDetailResponse> getDriverRideDetail(
            @Path("rideId") Long rideId
    );

    @GET("api/v1/rides/passenger")
    Call<ResponseBody> getPassengerRideHistory(
            @Query("startDate") String startDate,
            @Query("endDate") String endDate,
            @Query("page") int page,
            @Query("size") int size,
            @Query("sort") String sort
    );

    @GET("api/v1/rides/{rideId}/details/passenger")
    Call<PassengerRideDetailResponse> getPassengerRideDetail(
            @Path("rideId") Long rideId
    );

    @POST("api/v1/rides/{rideId}/rate")
    Call<ResponseBody> rateRide(
            @Path("rideId") Long rideId,
            @Body RideRatingRequest request
    );

    @GET("api/v1/rides/{rideId}/rating-status")
    Call<RideRatingStatusResponse> getRatingStatus(
            @Path("rideId") Long rideId
    );
  
    @PUT("api/v1/rides/{id}/cancel")
    Call<ResponseBody> cancelRide(@Path("id") Long rideId, @Body CancellationRequest request);

    @GET("api/v1/rides/{rideId}/track")
    Call<RideTrackingResponse> trackRide(
            @Path("rideId") Long rideId
    );

    @PUT("api/v1/rides/panic")
    Call<Void> panicRide(@Body PanicRideRequest request);

    @POST("api/v1/rides/{rideId}/report-inconsistency")
    Call<Void> reportInconsistency(
            @Path("rideId") Long rideId,
            @Body InconsistencyReportRequest request
    );

    @PUT("api/v1/rides/{rideId}/complete")
    Call<Void> completeRide(
            @Path("rideId") Long rideId
    );

    @PUT("api/v1/rides/stop")
    Call<StopRideResponse> stopRide(
            @Body StopRideRequest request
    );

    @PUT("api/v1/rides/{rideId}/vehicle-location")
    Call<Void> updateRideVehicleLocation(
            @Path("rideId") Long rideId,
            @Body UpdateVehicleLocationRequest request
    );
}